proposals/VK_EXT_external_memory_metal.adoc

```diff
@@ -0,0 +1,110 @@
+:refpage: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/
+:sectnums:
+
+This extension allows memory sharing through imports from or export to Metal resource handles.
+
+== Problem Statement
+
+An application may wish to reference device memory in multiple Vulkan device instances, in multiple processes, and/or in Metal API.
+This extension enables an application to export non-Vulkan handles from Vulkan memory objects such that the underlying resources can
+be referenced outside the scope of the Vulkan device instance that created them.
+
+While link:{refpage}VK_EXT_metal_objects.html[VK_EXT_metal_objects] provides a way to expose underlying Metal resources,
+when importing an image from an `id<MTLTexture>` handle, said images and their backing memory will be imported at
+link:{refpage}VkImage.html[VkImage] creation.
+According to point link:https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#resources-association[12.8 Memory Resource Association]
+resources are created as _virtual allocations_ with no backing memory. This leads to having to treat imported images through
+link:{refpage}VK_EXT_metal_objects.html[VK_EXT_metal_objects] differently.
+
+This extension aims to import both resources through link:{refpage}VkDeviceMemory.html[VkDeviceMemory] so that later can be bound
+to a existing link:{refpage}VkImage.html[VkImage] or link:{refpage}VkBuffer.html[VkBuffer] respectively.
+
+== Proposal
+
+This extension aims to standardize memory imports and exports of Vulkan resources using
+link:{refpage}VK_KHR_external_memory.html[VK_KHR_external_memory] as baseline to conform to what other platforms have done with
+extensions like link:{refpage}VK_KHR_external_memory_win32.html[VK_KHR_external_memory_win32] and
+link:{refpage}VK_ANDROID_external_memory_android_hardware_buffer.html[VK_ANDROID_external_memory_android_hardware_buffer] to name a few.
+
+This means that external Metal resources will be imported through link:{refpage}VkDeviceMemory.html[VkDeviceMemory] to later be bound.
+
+3 new bit values will be added to link:{refpage}VkExternalMemoryHandleTypeFlagBits.html[VkExternalMemoryHandleTypeFlagBits]:
+
+ - `VK_EXTERNAL_MEMORY_HANDLE_TYPE_MTLBUFFER_BIT_EXT` to accommodate for Metal buffers (MTLBuffer)
+ - `VK_EXTERNAL_MEMORY_HANDLE_TYPE_MTLTEXTURE_BIT_EXT` to accommodate for Metal textures (MTLTexture)
+ - `VK_EXTERNAL_MEMORY_HANDLE_TYPE_MTLHEAP_BIT_EXT` to accommodate for Metal heaps (MTLHeap)
+
+Due to MTLTexture's semantics in Metal API roughly align with that of an image with a dedicated allocation in Vulkan,
+implementations will be required to report `VK_EXTERNAL_MEMORY_FEATURE_DEDICATED_ONLY_BIT` for
+`VK_EXTERNAL_MEMORY_HANDLE_TYPE_MTLTEXTURE_BIT_EXT`. When importing a MTLTexture through the creation of a
+link:{refpage}VkDeviceMemory.html[VkDeviceMemory], the link:{refpage}VkImage.html[VkImage] it will be bound to
+must have the same creation parameters as the MTLTexture used at import.
```

Comment:
"must have same creation parameters as the MTLTexture used at import"

This needs a VU. A good example is VUID-VkMemoryAllocateInfo-pNext-02388, which is about AHardwareBuffer. There are more VUs in that neighborhood that are worth considering too. The new VU (new VUs) should enumerate which fields of VkImageCreateInfo must match the MTLTexture's creation params.

Reply:
Added a new VU to handle this case. I am thinking that maybe we could add a few more to force the users to provide the right size at allocation when importing buffers or heaps, but it kind of seems slightly pointless. Right now, we just say we don't really care about the size for buffers and heaps since we will use Metal to know the size of said resources. Any way, let me know your thoughts!

Reply:
I don't have a strong opinion on whether the spec should (a) require `size` to be correct or (b) ignore `size`. That's a decision is yours. In my opinion, the only practical difference is that choice (b) inhibits some VVL checks.

Regardless of the choice of (a) or (b) though for MTLBuffer and MTLHeap, more spec text is needed. If (a), then a VU is needed (similar to VUID-VkMemoryAllocateInfo-allocationSize-00647 or VUID-VkMemoryAllocateInfo-allocationSize-02383). If (b), then text similar to the MTLTexture "ignore" text is needed.

However, if you do choose (b), I prefer that the value not be actually ignored. In the past two years, several members of the Vulkan WG have become frustrated with the unintended consequences caused by ignored values in the spec. Superficially, it seems that an ignored `allocationSize` would cause no problems. But sometimes it can cause surprising difficulties for tools and layers. Instead of saying that `allocationSize` is ignored, I think it's better to add a VU that requires it to be 0.